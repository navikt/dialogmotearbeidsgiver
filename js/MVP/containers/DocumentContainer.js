import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { documentComponentPtMVP } from '../../propTypes';
import DocumentRenderer from '../components/DocumentRenderer';

const DocumentWrapperStyled = styled.div`
  display: flex;
  flex-direction: column;
  gap: 32px;
  border-radius: 4px;
  padding: 32px;
  background-color: white;
  margin-top: 32px;
`;

const DocumentContainer = ({ document, className, children }) => {
  return (
    <DocumentWrapperStyled className={className}>
      {document.map((documentComponent, index) => (
        <section key={index}>
          <DocumentRenderer documentComponent={documentComponent} />
        </section>
      ))}
      {children}
    </DocumentWrapperStyled>
  );
};

DocumentContainer.propTypes = {
  document: PropTypes.arrayOf(documentComponentPtMVP),
  className: PropTypes.string,
  children: PropTypes.node,
};

export default DocumentContainer;
