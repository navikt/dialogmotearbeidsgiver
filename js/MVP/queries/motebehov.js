import { useQuery } from 'react-query';
import { get } from '../../gateway-api';
import { getHentMotebehovUrl } from '../globals/paths';

const MOTEBEHOV = 'motebehov';

export const useMotebehov = (sykmeldt) => {
  const enabled = sykmeldt && !sykmeldt.isError && !sykmeldt.isLoading;

  const fnr = enabled ? sykmeldt.fnr : null;
  const virksomhetsnummer = enabled ? sykmeldt.orgnummer : null;

  return useQuery(
    [MOTEBEHOV, fnr, virksomhetsnummer],
    async () => {
      return get(getHentMotebehovUrl(fnr, virksomhetsnummer));
    },
    {
      enabled: !!enabled,
    }
  );
};
