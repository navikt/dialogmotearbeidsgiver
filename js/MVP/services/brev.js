import { get, post, getRaw } from '../../gateway-api/gatewayApi';
import { ISDIALOGMOTE_PROXY_BASE_PATH } from '../globals/paths';

export const postLestBrev = async (uuid) => {
  const url = `${ISDIALOGMOTE_PROXY_BASE_PATH}/${uuid}/les`;
  return post(url);
};

export const getBrev = async (fnr) => {
  const url = ISDIALOGMOTE_PROXY_BASE_PATH;
  const header = { 'nav-personident': fnr };
  return get(url, header);
};

export const getBrevPdf = async (uuid) => {
  const url = `${ISDIALOGMOTE_PROXY_BASE_PATH}/${uuid}/pdf`;
  return getRaw(url);
};
