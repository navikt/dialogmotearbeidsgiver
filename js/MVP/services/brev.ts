import { ISDIALOGMOTE_PROXY_BASE_PATH } from '../globals/paths';
import { get, post } from '@/api/axios';
import { Brev } from '@/api/types/brevTypes';

export const postLestBrev = (uuid: string): Promise<void> => {
  const url = `${ISDIALOGMOTE_PROXY_BASE_PATH}/${uuid}/les`;
  return post(url);
};

export const getBrev = (fnr?: string): Promise<Brev[]> => {
  return get(ISDIALOGMOTE_PROXY_BASE_PATH, { personIdent: fnr });
};

export const getBrevPdf = (uuid: string): Promise<Blob> => {
  const url = `${ISDIALOGMOTE_PROXY_BASE_PATH}/${uuid}/pdf`;
  return get(url, { responseType: 'blob' });
};
