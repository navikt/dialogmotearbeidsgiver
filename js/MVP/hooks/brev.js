import { useMutation, useQuery, useQueryClient } from 'react-query';
import { getBrev, postLestBrev } from '../services/brev';

const MOTEINNKALLELSE = 'moteinnkallelse';

export const useBrev = () => {
  return useQuery(MOTEINNKALLELSE, getBrev, {
    retry: 0,
  });
};

const setLestDatoForBrev = (uuid) => {
  return (varsel) => {
    if (varsel.uuid === uuid) {
      return { ...varsel, lestDato: true };
    }

    return varsel;
  };
};

export const useMutateBrevLest = () => {
  const queryClient = useQueryClient();

  return useMutation(({ uuid }) => postLestBrev(uuid), {
    onSuccess: (_data, variables) => {
      queryClient.setQueryData([MOTEINNKALLELSE], (old) => {
        return old.map(setLestDatoForBrev(variables.uuid));
      });
    },
  });
};
