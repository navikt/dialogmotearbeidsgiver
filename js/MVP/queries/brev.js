import { useMutation, useQuery, useQueryClient } from 'react-query';
import { getBrev, postLestBrev } from '../services/brev';
import { useSykmeldt } from './sykmeldt';

const brevQueryKeys = (koblingId) => ['brev', koblingId];

export const useBrev = (koblingId) => {
  const { fnr } = useSykmeldt(koblingId);
  return useQuery(brevQueryKeys(koblingId), () => getBrev(fnr), {
    enabled: !!fnr,
  });
};

const setLestDatoForBrev = (uuid) => {
  return (varsel) => {
    if (varsel.uuid === uuid) {
      return { ...varsel, lestDato: true };
    }

    return varsel;
  };
};

export const useMutateBrevLest = () => {
  const queryClient = useQueryClient();

  return useMutation(({ brevUuid }) => postLestBrev(brevUuid), {
    onMutate: (variables) => {
      queryClient.setQueryData(brevQueryKeys(variables.koblingId), (old) => {
        return old.map(setLestDatoForBrev(variables.brevUuid));
      });
    },
    onSettled: (variables) => {
      queryClient.invalidateQueries(brevQueryKeys(variables.koblingId));
    },
  });
};
