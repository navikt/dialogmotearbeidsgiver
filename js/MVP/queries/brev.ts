import { useMutation, useQuery, useQueryClient } from 'react-query';
import { getBrev, postLestBrev } from '../services/brev';
import { useSykmeldte } from './sykmeldte';

const brevQueryKeys = (narmestelederId) => ['brev', narmestelederId];

export const useBrev = (narmestelederId: string) => {
  const sykmeldteQuery = useSykmeldte(narmestelederId);
  const fnr = sykmeldteQuery.isSuccess ? sykmeldteQuery.data.fnr : null;
  return useQuery(brevQueryKeys(narmestelederId), () => getBrev(fnr), {
    enabled: sykmeldteQuery.isSuccess && !!sykmeldteQuery.data.fnr,
  });
};

const setLestDatoForBrev = (uuid: string) => {
  return (varsel) => {
    if (varsel.uuid === uuid) {
      return { ...varsel, lestDato: new Date().toString() };
    }

    return varsel;
  };
};

export const useMutateBrevLest = () => {
  const queryClient = useQueryClient();

  return useMutation(({ brevUuid }) => postLestBrev(brevUuid), {
    onMutate: (variables: { narmestelederId: string; brevUuid: string }) => {
      queryClient.setQueryData(brevQueryKeys(variables.narmestelederId), (old: string[]) => {
        return old.map(setLestDatoForBrev(variables.brevUuid));
      });
    },
    onSettled: (variables: { narmestelederId: string; brevUuid: string }) => {
      queryClient.invalidateQueries(brevQueryKeys(variables.narmestelederId));
    },
  });
};
