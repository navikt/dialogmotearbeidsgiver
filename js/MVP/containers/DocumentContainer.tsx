import { DocumentComponent } from '@/api/types/brevTypes';
import { useMutateBrevLest } from '@/MVP/queries/brev';
import React, { useEffect } from 'react';
import styled from 'styled-components';
import DocumentRenderer from '../components/DocumentRenderer';

const DocumentWrapperStyled = styled.div`
  display: flex;
  flex-direction: column;
  gap: 32px;
  border-radius: 4px;
  padding: 32px;
  background-color: white;
  margin-top: 32px;
  white-space: pre-wrap;
`;

interface DocumentContainerProps {
  document: DocumentComponent[];
  className?: string;
  children?: React.ReactNode;
  lestDato?: string;
  brevUuid: string;
}

const DocumentContainer = ({ document, lestDato, brevUuid, className, children }: DocumentContainerProps) => {
  const mutation = useMutateBrevLest();

  useEffect(() => {
    if (brevUuid && !lestDato && !mutation.isLoading) {
      mutation.mutate({ brevUuid });
    }
  }, [lestDato, mutation, brevUuid]);

  return (
    <DocumentWrapperStyled className={className}>
      {document.map((documentComponent, index) => (
        <section key={index}>
          <DocumentRenderer documentComponent={documentComponent} />
        </section>
      ))}
      {children}
    </DocumentWrapperStyled>
  );
};

export default DocumentContainer;
