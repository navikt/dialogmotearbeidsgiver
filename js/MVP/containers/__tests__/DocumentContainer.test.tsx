import { createDocumentComponent } from '@/MVP/test/fixtures/brev';
import { render } from '@testing-library/react';
import React from 'react';
import DocumentContainer from '../DocumentContainer';
import { useMutateBrevLest } from '@/MVP/queries/brev';

describe('DocumentContainer', () => {
  const mutateSpy = jest.fn();

  beforeEach(() => {
    (useMutateBrevLest as jest.Mock).mockReturnValue({
      mutate: mutateSpy,
      isLoading: false,
    });
  });

  afterEach(() => {
    mutateSpy.mockReset();
  });

  test('should render', () => {
    const document = [createDocumentComponent(), createDocumentComponent()];
    const { container } = render(
      <DocumentContainer document={document} brevUuid="brev_uuid">
        <div>Test div</div>
      </DocumentContainer>
    );

    expect(container).toHaveTextContent('Test_text_1');
  });

  test('should invoke mutation.mutate when lestDato is undefined', () => {
    const document = [createDocumentComponent(), createDocumentComponent()];
    render(
      <DocumentContainer document={document} brevUuid="brev_uuid">
        <div>Test div</div>
      </DocumentContainer>
    );

    expect(mutateSpy).toHaveBeenCalledWith({ brevUuid: 'brev_uuid' });
  });

  test('should not invoke mutation.mutate when lestDato is defined', () => {
    const document = [createDocumentComponent(), createDocumentComponent()];
    render(
      <DocumentContainer document={document} brevUuid="brev_uuid" lestDato="date">
        <div>Test div</div>
      </DocumentContainer>
    );

    expect(mutateSpy).not.toHaveBeenCalled();
  });
});
