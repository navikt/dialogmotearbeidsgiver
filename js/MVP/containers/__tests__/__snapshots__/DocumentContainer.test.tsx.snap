// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentContainer should render 1`] = `
<div>
  <div
    class="sc-bdvvtL bdogIK"
  >
    <section>
      <p
        class="typo-element"
      >
        TEST_HEADER
      </p>
      <p
        class="typo-normal"
      >
        Test_text_1
      </p>
    </section>
    <section>
      <p
        class="typo-element"
      >
        TEST_HEADER
      </p>
      <p
        class="typo-normal"
      >
        Test_text_1
      </p>
    </section>
    <div>
      Test div
    </div>
  </div>
</div>
`;
